### PR TITLE
Add application deployment information to the admin page

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Forem administrators are now able to to determine from the admin page (http://localhost:3000/admin) when they last deployed and what was the last commit ID that was deployed.

I used the following to determine how to set the deploy and commit variables based on environment:
|                  | Forem Cloud Variables | Heroku Variables          |
|------------------|-----------------------|---------------------------|
| Last deployed    | FOREM_BUILD_DATE      | HEROKU_RELEASE_CREATED_AT |
| Latest Commit ID | FOREM_BUILD_SHA       | HEROKU_SLUG_COMMIT        |

I want to note that I did not use the RELEASE_FOOTPRINT which differs from what the `data-release-footprint` key in the `<body>` tag uses. Instead I used the FOREM_BUILD_DATE and FOREM_BUILD_SHA which seemed more apt. I could use some validation on this, and perhaps we could change the data keys in the body to reflect the same info as the admin page.

## Related Tickets & Documents
https://github.com/forem/InternalProjectPlanning/issues/230
Related slack thread: https://forem-team.slack.com/archives/C0134R988P9/p1607586614240100

## QA Instructions, Screenshots, Recordings
Add export HEROKU_RELEASE_CREATED_AT="Optional" to the env file to set the HEROKU_RELEASE_CREATED_AT variable. 

The reason that it says `Optional` under the values is because as per above comment I've set those values in the env file with the string Optional.
Screenshot 2021-01-06 at 12 35 14

Besides changing the values for HEROKU_SLUG_COMMIT and HEROKU_RELEASE_CREATED_AT in the env file, I don't think there's a ton I can do to test this in development. Hence, I'd like to get the opinion of someone who is more knowledgeable than me on #systems and how these export variables are integrated into the app to ensure that this looks like it would work in prod. Based on this discussion I could put it behind a feature flag to test. 

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [ ] No documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: .env_sample
Comment (rhymes) [inline .env_sample:21]: Instead of using "Optional" for these two, can we keep the empty? This way you can test for presence and display "Not available" or something similar in case they aren't

**Review by rhymes — COMMENTED** [review]
A way to test write a test for this would be to inject environment variables as a setup for the system test, something like: `allow(ENV).to receive(:[]).with("VAULT_TOKEN").and_return("present")` (copied and pasted it from `lib/app_secrets_spec.rb`)

**Comment by Ridhwana** [comment]
> A way to test write a test for this would be to inject environment variables as a setup for the system test, something like: `allow(ENV).to receive(:[]).with("VAULT_TOKEN").and_return("present")` (copied and pasted it from `lib/app_secrets_spec.rb`)

@rhymes thanks! This sounds good - I was just waiting for a review that this is the right path before I worked on the tests.

File: .env_sample
Comment (Ridhwana) [reply .env_sample:21]: Sounds good, I'll change it 👍🏽

File: app/views/admin/admin_portals/index.html.erb
Comment (jdoss) [inline app/views/admin/admin_portals/index.html.erb:4]: `FOREM_BUILD_DATE` is the datetime when the container was created. When we start the container `RELEASE_FOOTPRINT` is the point in time that the container was started or deployed. I would change this to Container Build Date and make another section for Last Deployed that uses `RELEASE_FOOTPRINT` for the last time Forem was deployed.

**Review by jdoss — CHANGES_REQUESTED** [review]

File: app/views/admin/admin_portals/index.html.erb
Comment (citizen428) [inline app/views/admin/admin_portals/index.html.erb:5]: I'm not saying we need this now, BUT I think it could make sense to add a new model (not DB backed) like `ForemInstance` to which we add a bunch of class methods that encapsulate things like this:

```ruby
class ForemInstance
  def self.deployed_at
    ApplicationConfig["FOREM_BUILD_DATE"].presence || ENV["HEROKU_RELEASE_CREATED_AT"
  end

  def self.latest_commit_id
    ApplicationConfig["FOREM_BUILD_SHA"].presence || ENV["HEROKU_SLUG_COMMIT"]
  end
end
```

We probably will have a few more and I'd maybe even move `dev_to?` from `SiteConfig` to that new class since `ForemInstance.dev_to?` makes a lot more sense to me than `SiteConfig.dev_to?`. We could even move the `local?` method for the same reason (`ForemInstance.local?`). 

I found that in a lot of Rails apps the main concept itself often has no representation in code (e.g. blogs have `Author`, `Post` and `Comment` models, but nothing to represent the whole blog), but they often come in handy to collect certain utility functions in a sensible location.

File: app/views/admin/admin_portals/index.html.erb
Comment (Ridhwana) [reply app/views/admin/admin_portals/index.html.erb:4]: @jdoss thank you so much for this 😊 I'll use RELEASE_FOOTPRINT for Last deploy date. I don't think our users would need to know the Container Build Date so I'll leave that out. Thanks!

File: app/views/admin/admin_portals/index.html.erb
Comment (Ridhwana) [reply app/views/admin/admin_portals/index.html.erb]: @citizen428 this makes sense. I didn't include the changes in this PR, but I'm creating a follow up PR tomorrow. Thanks for the feedback :)

File: spec/requests/admin/admin_portals_spec.rb
Comment (rhymes) [inline spec/requests/admin/admin_portals_spec.rb:28]: non blocking: this sets the variable globally for any test

**Review by rhymes — APPROVED** [review]
LGTM

**Review by jdoss — APPROVED** [review]
LGTM!

File: spec/requests/admin/admin_portals_spec.rb
Comment (Ridhwana) [reply spec/requests/admin/admin_portals_spec.rb:28]: @rhymes I tried to set it so many other like using         
`allow(ENV).to receive(:[]).with("HEROKU_RELEASE_CREATED_AT").and_return("")` but I just couldn't get it to work. I'll look into it again tomorrow.

**Comment by Ridhwana** [comment]
Will look into why this test is failing...

File: spec/requests/admin/admin_portals_spec.rb
Comment (rhymes) [reply spec/requests/admin/admin_portals_spec.rb:28]: Ah got it! No worries, we can keep it like this as well :)

Comment (citizen428) [reply spec/requests/admin/admin_portals_spec.rb:28]: @Ridhwana Maybe something like this?

```ruby
stub_const("ENV", ENV.to_h.merge("HEROKU_RELEASE_CREATED_AT" => ""))
```

File: app/views/admin/admin_portals/index.html.erb
Comment (citizen428) [reply app/views/admin/admin_portals/index.html.erb]: No rush, like I said it really doesn't need to be part of this PR, but I feel like this will be a useful addition. Thanks for doing a follow-up PR! 😃

File: spec/requests/admin/admin_portals_spec.rb
Comment (Ridhwana) [reply spec/requests/admin/admin_portals_spec.rb:28]: Thanks @citizen428 this worked!

File: spec/helpers/application_helper_spec.rb
Comment (Ridhwana) [inline spec/helpers/application_helper_spec.rb:51]: This test was failing because it seems like ApplicationConfig["RELEASE_FOOTPRINT"] was "", hence it was returning early in the test based on this code 

```
    release_footprint = ApplicationConfig["RELEASE_FOOTPRINT"]
    return path if release_footprint.blank?

    "#{path}-#{params[:locale]}-#{release_footprint}-#{SiteConfig.admin_action_taken_at.rfc3339}"
```

I've reset the value in the test again. I went through my code and I'm not totally sure why this failed in this PR as I'm not resetting the value of RELEASE_FOOTPRINT anywhere in my tests, however it seems very coincidental. 

Whilst this is a fix for the test, I wanted a second pair of 👀's before merging in case I've overlooked something.  

cc @rhymes

Comment (rhymes) [reply spec/helpers/application_helper_spec.rb:51]: @Ridhwana I think it was failing because of `config/initializers/set_release_footprint.rb` which is initialized to `HEROKU_SLUG_COMMIT` on startup? That's the only clue I have :D

**Pull request merged**
